src/jwtMiddleware.ts

```diff
@@ -88,53 +89,57 @@ export const jwtMiddleware = <
 >({
   key,
   algorithm,
-  critHandlers,
   customMessages = {},
   ignorePatterns,
-  onSuccess,
-  onFailure,
+  onSuccess = () => {},
+  onFailure = () => true,
 }: JwtMiddlewareOptions): T => {
   Object.assign(customMessages, errorMessages);
 
   const core: RouterMiddleware = async (ctx, next) => {
-    if (!ignorePatterns || !ignorePath(ctx, ignorePatterns)) {
-      let isUnauthorized = true;
-      let validatedJwt;
-      if (ctx.request.headers.has("Authorization")) {
-        const authHeader = ctx.request.headers.get("Authorization")!;
-
-        if (authHeader.startsWith("Bearer ") && authHeader.length > 7) {
-          const jwt = authHeader.slice(7);
-
-          validatedJwt = await validateJwt({
-            jwt,
-            key,
-            algorithm,
-            critHandlers,
-          });
-
-          if (validatedJwt.isValid) {
-            isUnauthorized = false;
-            onSuccess && onSuccess(ctx, validatedJwt);
-          }
-        }
+    const onUnauthorized = async (
+      jwtValidation: Error,
+    ) => {
+      const shouldThrow = onFailure(ctx, jwtValidation);
+      if (shouldThrow) {
+        ctx.throw(
+          Status.Unauthorized,
+          customMessages?.ERROR_INVALID_AUTH,
+        );
       }
 
-      if (isUnauthorized) {
-        if (onFailure) {
-          const ignoreFailure = onFailure(ctx, validatedJwt);
-
-          if (!ignoreFailure) return;
-        } else {
-          ctx.throw(
-            Status.Unauthorized,
-            customMessages?.ERROR_INVALID_AUTH,
-          );
-        }
-      }
+      await next();
+    };
+
+    // If request matches ignore, call next early
+    if (ignorePatterns && ignorePath(ctx, ignorePatterns)) {
+      await next();
+
+      return;
     }
 
-    await next();
+    // No Authorization header
+    if (!ctx.request.headers.has("Authorization")) {
+      await onUnauthorized(new Error("Authorization header is not present"));
+
+      return;
+    }
+
+    // Authorization header has no Bearer or no token
+    const authHeader = ctx.request.headers.get("Authorization")!;
+    if (!authHeader.startsWith("Bearer ") || authHeader.length <= 7) {
+      await onUnauthorized(new Error("Invalid Authorization header"));
+
+      return;
+    }
+
+    const jwt = authHeader.slice(7);
+    try {
+      onSuccess(ctx, await verify(jwt, key, algorithm));
+      await next();
+    } catch (e) {
+      await onUnauthorized(e);
+    }
```

Comment:
Let's expand `onUnauthorized` with a second param (can be an optional boolean). Reason for this is that we can isolate the djwt errors from the middleware errors, and can return the djwt error instead of a predefined error as there is no easy way to distinguish the two. 

```suggestion
    try {
      onSuccess(ctx, await verify(jwt, key, algorithm));
      await next();
    } catch (e) {
      await onUnauthorized(e, true);
    }
```

Reply:
I tend not to be a big fan of passing flags as second arguments. Here I think it's not too bad since it's an internal thing so I'm fine with it 👍 

What type of different handling were you thinking of doing? Were you thinking of sending this flag into the `onFailure` callback, for instance?

Reply:
see the comment bellow